website/server/libs/payments/apple.js

```diff
@@ -194,48 +194,48 @@ api.noRenewSubscribe = async function noRenewSubscribe (options) {
     throw new NotAuthorized(api.constants.RESPONSE_NO_ITEM_PURCHASED);
   }
 
-  let transactionId;
+  let correctReceipt = false;
 
+  /* eslint-disable no-await-in-loop */
   for (const purchaseData of purchaseDataList) {
-    const dateTerminated = new Date(Number(purchaseData.expirationDate));
-    if (purchaseData.productId === sku && dateTerminated > new Date()) {
-      transactionId = purchaseData.transactionId;
+    if (purchaseData.productId === sku) {
+      const { transactionId } = purchaseData;
+      const existingReceipt = await IapPurchaseReceipt.findOne({
+        _id: transactionId,
+      }).exec();
+      if (existingReceipt) throw new NotAuthorized(this.constants.RESPONSE_ALREADY_USED);
+
+      await IapPurchaseReceipt.create({
+        _id: transactionId,
+        consumed: true,
+        // This should always be the buying user even for a gift.
+        userId: user._id,
+      });
+      const data = {
+        user,
+        paymentMethod: this.constants.PAYMENT_METHOD_APPLE,
+        headers,
+        sub,
+        autoRenews: false,
+      };
+
+      if (gift) {
+        gift.member = await User.findById(gift.uuid).exec();
+        gift.subscription = sub;
+        data.gift = gift;
+        data.paymentMethod = this.constants.PAYMENT_METHOD_GIFT;
+      }
+
+      await payments.createSubscription(data);
+      correctReceipt = true;
       break;
     }
   }
+  if (!correctReceipt) throw new NotAuthorized(api.constants.RESPONSE_INVALID_ITEM);
```

Comment:
I see that you changed the error code here from `INVALID_RECEIPT` to `INVALID_ITEM` -- is that more appropriate for the fail case if a correct receipt is not found in `purchaseDataList`?

Reply:
Yes, since this happens when the receipt itself is valid, but the contents are not correct (p.e. there is no purchase that matches the SKU provided)